doc/configuration.md
# Vizier Configuration

The Vizier web service API and all worker components are configured using environment variables in an attempt to follow [The Twelve-Factor App methodology](https://12factor.net/) for web application development. All relevant environment variables have default values that allow you to run the web service using a default configuration without the need to set any of the variables. The current default configuration will run the service in development mode.



## Web Service REST-API

The following environment variables control different parameters of the web API server. Additional variables to configure the different execution backends are described below.


### General

- ***VIZIERSERVER_NAME***: Web service name
- ***VIZIERSERVER_LOG_DIR***: Log file directory used by the web server (DEFAULT: *./.vizierdb/logs*)
- ***VIZIERSERVER_DEBUG***: Flag indicating whether server is started in debug mode (DEFAULT: *True*)


### Web Service

- ***VIZIERSERVER_BASE_URL***: Base URL of the server running the web service (DEFAULT: http://localhost)
- ***VIZIERSERVER_SERVER_PORT***: Public server port (DEFAULT: *5000*)
- ***VIZIERSERVER_SERVER_LOCAL_PORT***: Locally bound server port (DEFAULT: *5000*)
- ***VIZIERSERVER_APP_PATH***: Application path for Web API (DEFAULT: */vizier-db/api/v1*)
- ***VIZIERSERVER_ROW_LIMIT***: Default row limit for requests that read datasets (DEFAULT: *25*)
- ***VIZIERSERVER_MAX_ROW_LIMIT***: Maximum row limit for requests that read datasets (DEFAULT: *-1* (returns all rows))
- ***VIZIERSERVER_MAX_UPLOAD_SIZE***: Maximum size for file uploads in bytes (DEFAULT: *16777216*)

The distinction between *VIZIERSERVER_SERVER_PORT* and *VIZIERSERVER_SERVER_LOCAL_PORT* is relevant when running the web service inside a Docker container. Otherwise the value for both variables should be identical.


### Workflow Execution Engine

- ***VIZIERSERVER_ENGINE***: Name of the workflow execution engine (DEFAULT: *DEV*)
- ***VIZIERSERVER_PACKAGE_PATH***: Path to the package declaration directory (DEFAULT: *./resources/packages/common*)
- ***VIZIERSERVER_PROCESSOR_PATH***: Path to the task processor definitions for supported packages (DEFAULT: *./resources/processors/common:./resources/processors/dev*)


The files in the package directory are used to initialize the registry of supported packages. Each file in the packages directory is expected to contain the declaration of one or more vizier packages. All files are read during service initialization when repository of available workflow packages is created.



## Workflow Engine

The *vizier workflow engine* is used by the web API to create, delete, and manipulate projects and workflows. The engine is a wrapper for the four main system components: (1) the datastore for storing and manipulating datasets, (2) the filestore that manages uploaded files, (3) the viztrails repository for maintaining projects and workflows, and (4) the execution backend for running workflow modules. Different configurations for the engine can use different implementations of datastores, filestores, viztrails repositories, and execution backends. The engine configuration is set using the variable *VIZIERSERVER_ENGINE*. Vizier currently supports three different engine configurations:

- *DEV*: Development engine uses a simple datastore that maintains all datasets as separate files on the file system.
- *MIMIR*: The Mimir engine uses the Mimir gateway to store and manipulate datasets.
- *CLUSTER*: The cluster engine runs each project in an individual Docker container. In this configuration the engine itself does not contain instances of the datastore and filestore. Instead, each of the containers will have their own datastore, filestore, and execution backend.

At this point there exists only one implementation for the viztrails repository interface (*vizier.viztrails.objectstore*) as well as for the filestore interface (*vizier.filestore.fs*). Both implementations are therefore used by all three configurations.

The vizier engine is further configured using the following four environment variables:

- ***VIZIERENGINE_BACKEND***: Name of the execution backend. The currently implemented backends are CELERY, MULTIPROCESS, or CONTAINER (DEFAULT: MULTIPROCESS).
- ***VIZIERENGINE_SYNCHRONOUS***: Colon separated list of package.command strings that identify the commands that are executed synchronously (DEFAULT: None)
- ***VIZIERENGINE_USE_SHORT_IDENTIFIER***: Flag indicating whether short identifiers (eight characters instead of 32) are used by the viztrail repository (DEFAULT: True)
- ***VIZIERENGINE_DATA_DIR***: Base data directory for storing data. The datastore, filestore, and viztrail repository will create sub-folders in the directory for maintaining information and resources they maintain.

Each execution backend may use additional environment variables for its configuration. **Note** that not all combinations of engine configuration and backend name are valid. The backends *MULTIPROCESS* and *CELERY* can only be used in combination with engine configurations *DEV* and *MIMIR*. Backend *CONTAINER* is the backend when using engine configuration *CLUSTER*.


### MULTIPROCESS Backend

The MULTIPROCESS backend executes workflow modules using separate processes within the web service API. For each module a new process is created. This backend is primarily intended for installations on a local machine with a single user. There are no additional environment variables to configure the backend.


### CELERY Backend

The CELERY backend uses the distributed  task queue [Celery](http://www.celeryproject.org) to execute workflow modules using one or more Celery workers. This is a scalable configuration that is preferred over the MULTIPROCESS backend for installation with multiple users. The CELERY backend has the additional benefit that different modules can be executed by dedicated workers. This allows, for example, to execute Python cells in a notebook using a dedicated (containerized) worker to reduce the risk of malicious users interfering with the system. **Note**, however, that at this point workers are still shared by different users. To ensure complete isolation of user projects the *CLUSTER* configuration should be used.

The CELERY backend has two additional environment variables for its configuration:

- ***VIZIERENGINE_CELERY_ROUTES***: Colon separated list of package.command.queue strings that define routing information for individual commands (DEFAULT: None)
- ***CELERY_BROKER_URL***: URL for the celery broker (DEFAULT: amqp://guest@localhost//)

The default configuration uses [RabbitMQ](https://www.rabbitmq.com) as the message broker. Celery workers are configured using a set of environment variables as described below.


### CONTAINER Backend

When using the CONTAINER backend each vizier projects runs in a separate isolated container. This configuration is intended to prevent users from manipulating (or destroying) the projects of other users. The CONTAINER backend can only be used in combination with *CLUSTER* as the engine configuration. The CONTAINER backend is configured using two additional environment variables:

- ***VIZIERENGINE_CONTAINER_PORTS*** : List of port numbers for new project containers. Expects a comma-separated list of port number of number intervals (e.g. 8080-8088,9000,9090-9099)  (DEFAULT: 20171-20271)
- ***VIZIERENGINE_CONTAINER_IMAGE***: Unique identifier of the docker image for project containers (DEFAULT: *heikomueller/vizierapi:container*)

Each project container exposes a limited version of the web service API via a given port on the host machine. Port numbers are drawn from the list of number in *VIZIERENGINE_CONTAINER_PORTS*. It is assumed that all port numbers in the given list are available. Once all numbers are in use no new projects can be added. Port numbers are released when a project is deleted.



## Worker Configuration

Celery worker configuration is controlled by four environment variables:

- ***VIZIERWORKER_ENV***: Identifier for environment in which the worker operates (supported values are *DEV*, *MIMIR*, and *REMOTE*) (DEFAULT: *DEV*)
- ***VIZIERWORKER_PROCESSOR_PATH***: Path to the task processor definitions for supported packages (DEFAULT: *./resources/processors/common:./resources/processors/dev*)
- ***VIZIERWORKER_LOG_DIR***: Log file directory used by the worker (DEFAULT: *./.vizierdb/logs/worker*)
- ***VIZIERWORKER_CONTROLLER_URL***: URL of the controlling web service (DEFAULT: http://localhost:5000/vizier-db/api/v1)

 In addition, the variables *CELERY_BROKER_URL* and *VIZIERENGINE_DATA_DIR* are also used by the workers.

The value of the environment variable *VIZIERWORKER_ENV* should either match the value of *VIZIERSERVER_ENGINE* or be *REMOTE*. The remote case is intended for running dedicated workers that execute Python cells. In a remote environment the worker will use the remote datastore client to read and write datasets. Thus, the worker does not need access to the local file system and can be run in an isolated container. The remote datastore client is initialized using the same URL that is used by the worker controller (set in *VIZIERWORKER_CONTROLLER_URL*).



## Packages and Task Processors

The list of available commands that can be executed as workflow modules (i.e., notebook cells) is defined using the files in the in directories in *VIZIERSERVER_PACKAGE_PATH*. The path is a colon-separated list of local directories. Every file in each of the directories is expected to contain a package declaration. See [Packages in Vizier](https://github.com/VizierDB/web-api-async/blob/master/doc/packages.md) for more details on the file format. Declarations for common packages can be found in the directory [resources/packages/common](https://github.com/VizierDB/web-api-async/tree/master/resources/packages/common/). Declarations for additional packages that are only available when running the Mimir configurations can be found in the directory  [resources/packages/mimir](https://github.com/VizierDB/web-api-async/tree/master/resources/packages/mimir/).

For each package a task processor needs to be specified to execute the commands that are defined in the package. Task processors should implement the interface [TaskProcessor](https://github.com/VizierDB/web-api-async/blob/master/vizier/engine/task.processor.py). Task processors are instantiated from files that are found in the directories in the *VIZIERSERVER_PROCESSOR_PATH* (or *VIZIERWORKER_PROCESSOR_PATH* for Celery workers). The expected file format is is:

```
packages:
    - '...' # List of package identifier
engine:
    className: '...' # Name of the Python class that implements the task processor
    moduleName: '...' # Name of the module that contains the task processor class
    properties: # Optional dictionary of processor-specific parameters
        ...
```

Files can either be serialized as JSON or Yaml. Task processor definitions for the common packages can be found in the directory [resources/processors/common](https://github.com/VizierDB/web-api-async/tree/master/resources/processors/common/). The task processor for VizUAL commands when running hte development configuration is found in [resources/processors/dev](https://github.com/VizierDB/web-api-async/tree/master/resources/processors/dev/) while task processors for additional packages in the Mimir configuration are maintained in directory [resources/processors/mimir](https://github.com/VizierDB/web-api-async/tree/master/resources/processors/mimir/).

### Task Processor for Plot Package

The Yaml serialization that instantiates the task processor for plot commands is:

```
packages:
    - plot
engine:
    className: 'PlotProcessor'
    moduleName: 'vizier.engine.packages.plot.processor'
```

The plot task processor does not take any additional arguments.

### Task Processor for VizUAL Package

The task processor for VizUAL commands takes an additional argument *api* that specifies the Python class that implements the VizUAL API that is used by the task processor (depending on whether *DEV* or *MIMIR* is used as the engine configuration).

**VizUAL Task Processor for DEV Engine**

```
packages:
    - vizual
engine:
    className: 'VizualTaskProcessor'
    moduleName: 'vizier.engine.packages.vizual.processor'
    properties:
        api:
            className: 'DefaultVizualApi'
            moduleName: 'vizier.engine.packages.vizual.api.fs'
```

**VizUAL Task Processor for MIMIR Engine**

```
packages:
    - vizual
engine:
    className: 'VizualTaskProcessor'
    moduleName: 'vizier.engine.packages.vizual.processor'
    properties:
        api:
            className: 'MimirVizualApi'
            moduleName: 'vizier.engine.packages.vizual.api.mimir'
```


changelog.md
# Vizier Web API - Changelog

### 0.1.0

* Initial Version

### 0.2.0

* Second prototype

### 0.3.0 - 2018-01-22

* Workflow branches
* Dataset annotations
* Drop and rename datasets
* Mimir lenses and data store
* Deletion of workflow modules
* Access full history of workflows
* Access repository of available workflow modules via API
* Allow different workflow engines for different projects

### 0.3.1 - 2018-02-08

* Replace workflow engine dictionary with dictionary of execution environments
* Change workflow context to be a dictionary


### 0.3.2 - 2018-03-05

* Add file upload size limit to service description
* Enable upload of files that do not parse correctly as CSV/TSV files


### 0.4.0 - 2018-05-23

* Add support for dataset chart views
* Fix issues with dataset upload and column data type awareness
* Pagination for datasets
* Switch to column ids in module arguments
* Server-side command text for workflow modules
* Default label for data series
* Enable multiple annotation with same key for single object (e.g., multiple user comments)


### 0.4.1 - 2018-06-07

* Add GEOCODE lens


### 0.4.2 - 2018-06-12

* Push pagination queries to datastore
* Avoid materializing Mimir annotations in dataset metadata
* Encode special characters as unicode in dataset creation


### 0.4.3 - 2018-06-13

* Add VizUAL Sort and Filter Columns commands


### 0.4.4 - 2018-06-18

* Upload files from URL


### 0.5.0 - 2019-01-23

This is the first prototype that supports asynchronous execution of workflow modules. The prototype was created from the code base of the web API version 0.4.4. It incorporates changes that were made to the UB branch and that were merged into the master branch while the prototype was under development.

* Change definition and handling of packages
* Allow concurrent execution of workflow modules
* Datastore that uses the web API to access and manipulate dataset instead of the file system
* Support containerization of individual projects
* Command Line Interface to interact with the web API


vizier/config.py
# Copyright (C) 2018 New York University
#                    University at Buffalo,
#                    Illinois Institute of Technology.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Application configuration object. Contains all application settings and the
list of available packages.

The app configuration has three main parts: API and Web service configuration,
list of available packages, and the workflow execution environment. The
structure of the configuration object is as follows:

webservice:
    server_url: Url of the server (e.g., http://localhost)
    server_port: Public server port (e.g., 5000)
    server_local_port: Locally bound server port (e.g., 5000)
    app_path: Application path for Web API (e.g., /vizier-db/api/v1)
    app_base_url: Concatenation of server_url, server_port and app_path
    doc_url: Url to API documentation
    name: Web Service name
    defaults:
        row_limit: Default row limit for requests that read datasets
        max_row_limit: Maximum row limit for requests that read datasets (-1 = all)
        max_file_size: Maximum size for file uploads (in byte)
packages: # List of package declarations
    - declarations: File containing declaration of package commands
      parameters: Package specific configuration parameters
engine:
    moduleName: Name of the module containing the used engine
    className: Class name of the used engine
    properties: Dictionary of engine specific configuration properties
debug: Flag indicating whether server is started in debug mode
logs:
    server: Log file for Web Server
"""

import json
import os
import yaml

from vizier.core.io.base import PARA_LONG_IDENTIFIER
from vizier.core.loader import ClassLoader
from vizier.engine.packages.base import PackageIndex
from vizier.engine.packages.vizual.base import PACKAGE_VIZUAL, VIZUAL_COMMANDS
from vizier.engine.packages.vizual.processor import PROPERTY_API

import vizier.datastore.fs.base as ds
import vizier.filestore.fs.base as fs
import vizier.viztrail.driver.objectstore.repository as vt


"""Environment Variable containing path to config file."""
ENV_CONFIG = 'VIZIERSERVER_CONFIG'

"""Default directory for API data."""
ENV_DIRECTORY = './.vizierdb'

"""Default settings."""
DEFAULT_SETTINGS = {
    'webservice': {
        'server_url': 'http://localhost',
        'server_port': 5000,
        'server_local_port': 5000,
        'app_path': '/vizier-db/api/v1',
        'doc_url': 'http://cds-swg1.cims.nyu.edu/doc/vizier-db/',
        'name': 'Vizier Web API',
        'defaults': {
            'row_limit': 25,
            'max_row_limit': -1,
            'max_file_size': 16 * 1024 * 1024
        }
    },
    'engine' : ClassLoader.to_dict(
        module_name='vizier.engine.environments.local',
        class_name='DefaultLocalEngine',
        properties={
            'datastore': ClassLoader.to_dict(
                module_name='vizier.datastore.fs.factory',
                class_name='FileSystemDatastoreFactory',
                properties={
                    ds.PARA_DIRECTORY: os.path.join(ENV_DIRECTORY, 'ds')
                }
            ),
            'filestore': ClassLoader.to_dict(
                module_name='vizier.filestore.fs.factory',
                class_name='DefaultFilestoreFactory',
                properties={
                    fs.PARA_DIRECTORY: os.path.join(ENV_DIRECTORY, 'fs')
                }
            ),
            'viztrails': ClassLoader.to_dict(
                module_name='vizier.viztrail.driver.objectstore.repository',
                class_name='OSViztrailRepository',
                properties={
                    vt.PARA_DIRECTORY: os.path.join(ENV_DIRECTORY, 'vt'),
                    PARA_LONG_IDENTIFIER: False
                }
            )
        }
    ),
    'logs': {
        'server': os.path.join(ENV_DIRECTORY, 'logs')
    },
    'debug': True,
}


class AppConfig(object):
    """Application configuration object. This object contains all configuration
    settings for the Vizier instance.

    Configuration is read from a configuration file (in Yaml or Json format).
    The file is expected to contain follow the structure of the configuration
    object as described above. The configuration file can either be specified
    using the environment variable VIZIERSERVER_CONFIG or be located (as file
    config.yaml) in the current working directory.
    """
    def __init__(self, configuration_file=None):
        """Read configuration parameters from a configuration file. The file is
        expected to be in Yaml format, having the same structure as shown above.

        If no file is specified attempts will be made to read the following
        configuration files:

        (1) The file referenced by the environment variable VIZIERSERVER_CONFIG
        (2) The file config.yaml in the current working directory

        If the specified files are not found the default configuration is used.

        Parameters
        ----------
        configuration_file: string, optional
            Optional path to configuration file
        """
        # Read configuration from given configuration file or one of the
        # options: value oin environment variable, local config.json file, or
        # local config.yaml file. Use the default settings if no configuration
        # file is found.
        doc = None
        if configuration_file != None:
            if os.path.isfile(configuration_file):
                doc = read_object_from_file(configuration_file)
            else:
                raise ValueError('unknown file \'' + str(configuration_file) + '\'')
        else:
            files = [
                os.getenv(ENV_CONFIG),
                './config.json',
                './config.yaml'
            ]
            for config_file in files:
                if not config_file is None and os.path.isfile(config_file):
                    doc = read_object_from_file(config_file)
                    break
        if doc is None:
            doc = DEFAULT_SETTINGS
        # Registry of available packages. By default, only the VizUAL package is
        # pre-loaded. Additional packages are loaded from files specified in the
        # 'packages' element (optional). Note that the list may contain a
        # reference to a VizUAL package which overrides the default package
        # declaration.
        self.packages = {
            PACKAGE_VIZUAL: PackageIndex(
                package=VIZUAL_COMMANDS,
                properties={
                    PROPERTY_API: ClassLoader.to_dict(
                        module_name='vizier.engine.packages.vizual.api.fs',
                        class_name='DefaultVizualApi'
                    )
                }
            )
        }
        if 'packages' in doc:
            for pckg_file in doc['packages']:
                if not 'declarations' in pckg_file:
                    raise ValueError('missing key \'declarations\' for package')
                pckg = read_object_from_file(pckg_file['declarations'])
                for key in pckg:
                    if 'parameters' in pckg_file:
                        package_parameters = pckg_file['parameters']
                    else:
                        package_parameters = None
                    self.packages[key] = PackageIndex(
                        package=pckg[key],
                        properties=package_parameters
                    )
        # Create a class loader instance for the engine.
        if 'engine' in doc:
            engine = ClassLoader(values=doc['engine'])
        else:
            engine = ClassLoader(values=DEFAULT_SETTINGS['engine'])
        setattr(self, 'engine', engine)
        # Create objects for configuration of log files and web service.
        for key in ['logs', 'webservice']:
            if key in doc:
                obj = ConfigObject(
                    values=doc[key],
                    default_values=DEFAULT_SETTINGS[key]
                )
            else:
                obj = ConfigObject(values=DEFAULT_SETTINGS[key])
            setattr(self, key, obj)
        # Set debug flag if given
        if 'debug' in doc:
            self.debug = doc['debug']
        else:
            self.debug = DEFAULT_SETTINGS['debug']

    @property
    def app_base_url(self):
        """Full Url (prefix) for all resources that are available on the Web
        Service. this is a concatenation of the service Url, port number and
        application path.

        Returns
        -------
        string
        """
        base_url = self.webservice.server_url
        if self.webservice.server_port != 80:
            base_url += ':' + str(self.webservice.server_port)
        base_url += self.webservice.app_path
        return base_url

    def get_api_engine(self):
        """Get the instance of the defined API engine.

        Returns
        -------
        vizier.engine.base.VizierEngine
        """
        return self.engine.get_instance(packages=self.packages)


class ConfigObject(object):
    """Create a Python class instance from a given dictionary of variables and
    a dictionary of default values. Performs a nested merge of dictionaries in
    values and default values.

    For lists the policy is all or nothing. List elements are not merged. If the
    configuration object contains a list value all items in the list are copied
    to the configuration as is.
    """
    def __init__(self, values, default_values=dict()):
        """Initialize the object from the given dictionary of values and default
        values.

        Parameters
        ----------
        values: dict
            Dictionary of object variables. Dictionary keys are used as variable
            names.
        default_values: dict
            Dictionary of default values.
        """
        #print json.dumps(values, sort_keys=True, indent=4)
        #print json.dumps(default_values, sort_keys=True, indent=4)
        self.keys = set()
        for key in values:
            if isinstance(values[key], dict):
                if key in default_values:
                    def_val = default_values[key]
                    if not isinstance(def_val, dict):
                        def_val = dict()
                else:
                    def_val = dict()
                setattr(
                    self,
                    key,
                     ConfigObject(values[key], default_values=def_val)
                )
            elif isinstance(values[key], list):
                elements = list()
                for el in values[key]:
                    if isinstance(el, dict):
                        elements.append(ConfigObject(el))
                    else:
                        elements.append(el)
                setattr(self, key,  elements)
            else:
                setattr(self, key,  values[key])
            self.keys.add(key)
        for key in default_values:
            if not key in self.keys:
                if isinstance(default_values[key], dict):
                    setattr(
                        self,
                        key,
                        ConfigObject(dict(), default_values=default_values[key])
                    )
                elif isinstance(default_values[key], list):
                    elements = list()
                    for el in default_values[key]:
                        if isinstance(el, dict):
                            elements.append(ConfigObject(el))
                        else:
                            elements.append(el)
                    setattr(self, key,  elements)
                else:
                    setattr(self, key,  default_values[key])
                self.keys.add(key)


# ------------------------------------------------------------------------------
# Helper Methods
# ------------------------------------------------------------------------------

def read_object_from_file(filename):
    """Read dictionary serialization from file. The file format is expected to
    by Yaml unless the filename ends with .json.

    Parameters
    ----------
    filename: string
        Name of the input file

    Returns
    -------
    dict
    """
    with open(filename, 'r') as f:
        if filename.endswith('.json'):
            return json.loads(f.read())
        else:
            return yaml.load(f.read())
